### DISCUSSION TITLE
How to get the rawBody in a custom endpoint

### DISCUSSION THREAD

**Discussion by saptha** [directus/directus#8633] [Archive]
### Preflight Checklist

- [X] I have completed all [Troubleshooting Steps](https://docs.directus.io/getting-started/support/#troubleshooting-steps).
- [X] I'm on [the latest version of Directus](https://github.com/directus/directus/releases).
- [X] There's [no other issue](https://github.com/directus/directus/issues) that already describes my problem.

### Describe the Bug

I create a custom endpoint which listens to events from Stripe payment gateway.
To verify the signature of the events from Stripe, I need to access the rawBody of the payload.
I couldn't find it in the documentation as to how to retrieve the raw input.
Kindly suggest.

### To Reproduce

I followed the steps as mentioned in the below Stripe docs:
https://stripe.com/docs/payments/checkout/fulfill-orders
https://stripe.com/docs/webhooks/signatures

### What version of Directus are you using?

9.0.0-rc.96

### What version of Node.js are you using?

v16.10.0

### What database are you using?

MySQL

### What browser are you using?

Chrome

### What operating system are you using?

macOS

### How are you deploying Directus?

running locally

**Comment by azrikahar** [comment]
_Moved this to discussion as it seems like more of a how-to-implement rather than a bug, but will move it back if it turns out to be an actual bug with Directus_.

> I create a custom endpoint which listens to events from Stripe payment gateway.

Can you share the source code for this custom endpoint?

> To verify the signature of the events from Stripe, I need to access the rawBody of the payload.
> I couldn't find it in the documentation as to how to retrieve the raw input.

As the links you shared suggest, the examples are shown in Express so that's great since Directus uses Express as well. 

Note that in the example given by Stripe, the payload you were mentioning is retrieved from `request.body` as seen here:

![chrome_J1C4urGugL](https://user-images.githubusercontent.com/42867097/136510500-a3a21349-705f-45ed-88f3-814e5e1a48a5.png)

In Directus custom endpoint, you will have access to the same `request` (and `.body` in it). Here's a barebone example:

```js
module.exports = function registerEndpoint(router, context) {
	router.post('/', (req, res, next) => {
		const payload = req.body; // "req" is short for "request" here
	});
};
```

Definitely recommend checking out the Directus custom API guide here for more info regarding what I showed above: https://docs.directus.io/guides/api-endpoints/, and feel free to let us know if the guides/docs aren't clear enough as we are trying to improve it as much as possible.

**Reply by saptha** [reply]
Exactly, I followed as per the Directus docs. But the API is unable to verify the signature of Stripe events. Throwing the below error in the console: 

> Stripe webhook failed with Error: No signatures found matching the expected signature for payload. Are you passing the raw request body you received from Stripe? https://github.com/stripe/stripe-node#webhook-signing

Here is my code for your reference

```js
module.exports = function registerEndpoint(router, { services, exceptions }) {
	const stripe = require('stripe')(process.env.STRIPE_TST_SECRET);
	const { PayloadService } = services;
	const { ForbiddenException } = exceptions;
        const bodyParser = require('body-parser');


	const fulfillOrder = (session) => {
		// TODO: fill me in
		console.log('Fulfilling order', session);
	};

	const createOrder = (session) => {
		// TODO: fill me in
		console.log('Creating order', session);
	};

	const emailCustomerAboutFailedPayment = (session) => {
		// TODO: fill me in
		console.log('Emailing customer', session);
	};

	router.post('/', bodyParser.raw({type: 'application/json'}), async (req, res) => {
		let event;
		const endpointSecret = process.env.STRIPE_TST_CLI_SECRET;
		const sig = req.headers['stripe-signature'];

		if (endpointSecret) {
			try {
				event = stripe.webhooks.constructEvent(req.body, sig, endpointSecret);
			} catch (err) {
				console.log(`Stripe webhook failed with ${err}`);
				return res.status(400).send(`Webhook Error: ${err.message}`);
			}
		}

		// Handle the event
		switch (event.type) {
			case 'checkout.session.completed': {
				const session = event.data.object;
				// Save an order in your database, marked as 'awaiting payment'
				createOrder(session);

				// Check if the order is paid (e.g., from a card payment)
				//
				// A delayed notification payment will have an `unpaid` status, as
				// you're still waiting for funds to be transferred from the customer's
				// account.
				if (session.payment_status === 'paid') {
					fulfillOrder(session);
				}

				break;
			}

			case 'checkout.session.async_payment_succeeded': {
				const session = event.data.object;

				// Fulfill the purchase...
				fulfillOrder(session);

				break;
			}

			case 'checkout.session.async_payment_failed': {
				const session = event.data.object;

				// Send an email to the customer asking them to retry their order
				emailCustomerAboutFailedPayment(session);

				break;
			}
		}

		res.status(200).end();
	});
};
```

**Reply by azrikahar** [reply]
```js
const sig = req.headers['stripe-signature'];
```

Can you verify whether this variable `sig` has the header value? Tracing whether the headers is properly received might be the key here.

I'd also recommend using the Stripe example as-is, meaning creating a whole new Express project to test it out, as it'll be easier to debug imo.

**Reply by saptha** [reply]
Yes, the `sig` has the header value in it.

As you suggested, I created it as a separate Express project with the Stripe example as-is and it works fine. No issues reported. 

I think I'll continue with that approach till this issue is fixed in Directus

Thanks for your help and support.

**Reply by azrikahar** [reply]
> Yes, the `sig` has the header value in it.

Thanks for verifying. 👍

I didn't notice this in your code earlier, but would you mind to try the following in the custom endpoint:

1. remove this line

   ```diff
   -  const bodyParser = require('body-parser');
   ```

2. change this line

   ```diff
   - router.post('/', bodyParser.raw({type: 'application/json'}), async (req, res) => {
   + router.post('/', (req, res) => {
   ```
 
---
  
If that still doesn't work, then it might be indeed unsupported currently due to this middleware:

https://github.com/directus/directus/blob/fc87c0b662903214ca301e0a6b4af21efe135422/api/src/app.ts#L95-L97

and we'll need to modify it as described in these references:

- https://stackoverflow.com/questions/53899365/stripe-error-no-signatures-found-matching-the-expected-signature-for-payload
- https://flaviocopes.com/express-get-raw-body/

**Reply by joselcvarela** [reply]
Hello @saptha 
You should need something like this to use stripe webhooks:
```js
    const sig = req.headers['stripe-signature'];

    const event = await stripe.webhooks.constructEvent(
      req.rawBody,
      sig,
      secret
    );
```

Am I right?

But the problem you are facing is, how to achieve the `req.rawBody`.
My suggestion is to add the 'middlewares.init.before' custom hook and attach this middleware to Express (not tested):
```js
// extensions/hooks/raw-body/index.js:

module.exports = function registerHook({}) {
  return {
    "middlewares.init.before": async function ({ app }) {
      app.use(function (req, res, next) {
        var data = "";
        req.on("data", function (chunk) {
          data += chunk;
        });
        req.on("end", function () {
          req.rawBody = data;
          next();
        });
      });
    },
  };
};
```

**Reply by saptha** [reply]
Thanks for the suggestion to use 'middlewares.init.before'  @joselcvarela , 
I used it but your sample code didnt work. The application hanged without returning anything. 
I got it working with the below code.
```

// extensions/hooks/raw-body/index.js:

module.exports = function registerHook({}) {
    const bodyParser = require('body-parser');

    return {
      "middlewares.init.before": async function ({ app }) {
        app.use(bodyParser.json({
            verify: (req, res, buf) => {
              req.rawBody = buf
            }
          }))
      },
    };
  };


```

**Comment by saptha** [comment]
Solution is to use `middlewares.init.before` hook and to attach rawBody to the incoming request.

```
// extensions/hooks/raw-body/index.js:

module.exports = function registerHook({}) {
    const express = require('express');

    return {
      "middlewares.init.before": async function ({ app }) {
        app.use(express.json({
            verify: (req, res, buf) => {
              if (req.originalUrl.startsWith('/webhook')) {
                req.rawBody = buf.toString()
              }
            }
          }))
      },
    };
  };
```

**Comment by eikaramba** [comment]
for anyone coming along the interface has slightly changed. this is working in directus 9.3.0
```js
const express = require('express');

module.exports = function registerHook({init}) {
  init("middlewares.before",async function ({ app }) {
        app.use(express.json({
	        verify: (req, res, buf) => {
	          if (req.originalUrl.startsWith('/path/to/your/webhook/endpoint'))
		          req.rawBody = buf.toString()
	        }
      }))
  })
};
```

**Reply by ph-arm** [reply]
I am having a similar issue with application/x-www-form-urlencoded body, just wondering if it should go with the solution type and create a hook.

Would be so cool to have some doc on this. Again might be a newbie question here

**Reply by ph-arm** [reply]
The hook work (I can log the raw body from , but it doesn't give me access to req.rawBody in the endpoint context (log undefined).
is there any neat solutions on that topic?

**Reply by eikaramba** [reply]
please share some code. i am not sure if i can otherwise help, as it should work. you should be able to find print req.rawBody to console. if log is undefined than it seems your logger is not initialized.

**Reply by ph-arm** [reply]
sorry if I was not clear.
Let me detail it below
the hook code before build :
```javascript
const express = require('express');
module.exports =  function registerHook({init}) {
  console.log("registering rawBody hook")
  init("middlewares.before",async function ({ app }) {
        app.use(express.json({
	        verify: (req, res, buf) => {
                   console.log("hook running")
                   console.log(req.originalUrl)
	           if (req.originalUrl.startsWith('/twilio-endpoint/callbacksSMS')){
		        console.log("transplitation of the body")
                        req.rawBody = buf.toString()
                        console.log("from hook: "+req.rawBody)
                   }
	        }
      }))
  })
};
```

and the endpoint code before build:
```javascript
[...]
 router.post("/callbacksSMS", async (req, res, next) => {
    const content = {};
    console.log(req.rawBody)
    let parsedData = await querystring.decode(req.rawBody)
    console.log(parsedData)
    [...]
    res.send(200);
  });
};
```
and got this as log :
```
22:22:01 ✨ request completed OPTIONS 204 /twilio-endpoint/sms 1ms
hook running
/twilio-endpoint/sms
express deprecated res.send(status): Use res.sendStatus(status) instead extensions/endpoints/twilio-endpoint/index.js:1:3221
22:22:01 ✨ request completed POST 200 /twilio-endpoint/sms 284ms
22:22:02 ✨ request completed GET 200 /assets/6dcf41a9-eb16-442a-b925-7165ca7d1dbd?width=24&height=24&fit=cover 11.7s
undefined
[Object: null prototype] {}
express deprecated res.send(status): Use res.sendStatus(status) instead extensions/endpoints/twilio-endpoint/index.js:1:3584
22:22:04 ✨ request completed POST 200 /twilio-endpoint/callbacksSMS 672ms
22:22:04 ✨ request completed GET 200 /assets/53b7de4c-eeeb-4a59-8b56-fd1f98ff52e8?width=24&height=24&fit=cover 14.4s
undefined
[Object: null prototype] {}
22:22:08 ✨ request completed POST 200 /twilio-endpoint/callbacksSMS 312ms
```
version is the latest 1.0.0
is there anything I missed? or not got it. Sorry to bother

**Reply by eikaramba** [reply]
it seems the rawBody is not set. it seems the if condition is not even true. You call `/twilio-endpoint/sms` but in your code you check for  `/twilio-endpoint/callbacksSMS`

**Reply by ph-arm** [reply]
I don t really understand your response. 
From the log, I don't seem to be running the hook at all on the POST, only on the Options/GET requests.
If I understand correctly, how it works, it should log "hook running" for every request, but here it just doesn't seem to run at all on the POST requests.
Content type from the post query is application/x-www-form-urlencoded. I can't change it since, it comes from twilio directly.

**Reply by eikaramba** [reply]
strange, for me it is working fine. for every request  the middleware is executed. maybe remove your router.post method altogehter and make sure the middleware is first executed every time. if not then something is off there.

**Reply by ph-arm** [reply]
I can confirm, checked this morning, the hook is not running on POST requests. (this is version 9.6 for info, updated today too, in case it could help , my mistake, because of the relooking )

**Reply by eikaramba** [reply]
what version is 1.0.0 ? directus current version is [9.7.1](https://github.com/directus/directus/releases/tag/v9.7.1)

**Comment by lucameusburger** [comment]
is there an update on this topic? i am running into the same issue when tryhing
`stripe.webhooks.constructEvent(req.body, sig, endpointSecret)`

and i tried to append the rawBody to the request via
```
import express, { NextFunction, Request, Response } from 'express';

import { defineHook } from '@directus/extensions-sdk';

export default defineHook(({ init }) => {
    init('middlewares.before', ({ app }) => {
        app.use((req: Request, res: Response, next: NextFunction) => {
            if (req.path === '/endpoint-tickets/webhook') {
                const chunks: Buffer[] = [];
                req.on('data', (chunk) => chunks.push(chunk));
                req.on('end', () => {
                    const rawBody = Buffer.concat(chunks);
                    (req as any).rawBody = rawBody;
                    next();
                });
            } else {
                next();
            }
        });
    });
});
```

but this then just throws an 400 error on the endpoint.

**Reply by eikaramba** [reply]
can you check this? https://github.com/eikaramba/directus-extension-stripe/blob/main/src/stripe-hook/index.ts#L77
i'm on mobile currently, but once you successfully adapted your code let us know :)

**Reply by lucameusburger** [reply]
This works... Kind of. I got the middlewares.before hook to work before and i just wanted to catch if the request was sent to my stripe webhook endpoint and then alter the body to be raw or append the rawBody somehow. but this did not work and also does not with your provided code. when i try to do this i get an 400 error from the endpoint itself.

the way i solved it now is just to move the code from the endpoint all into the middleware.before hook (as you did also). this works fine but i think is not the cleanest solution as this endpoint file has some shared code in it (for example the pdf generation) and i would have liked to have the webhook also in there to use this shared code. anyways it is a proper solution. i just think it would be cleaner if we just could get the rawBody to the endpoint itself.

the oter solution was actually to handle this in the nextjs of my website which also worked fine but i want to have everything in directus if possible.

But solid code of yours! Thanks!

here would be the nextjs route.ts that then verified the stripe hook and forwarded to my directus endpoint
```
import { NextResponse } from 'next/server';
import Stripe from 'stripe';

const stripe = new Stripe(process.env.STRIPE_SECRET_KEY!, {
    apiVersion: '2025-03-31.basil',
});

const endpointSecret = process.env.STRIPE_WEBHOOK_SECRET!;
const DIRECTUS_WEBHOOK_URL = process.env.DIRECTUS_WEBHOOK_URL!;

export async function POST(request: Request) {
    const body = await request.text();
    const signature = request.headers.get('stripe-signature');

    if (!signature || !endpointSecret) {
        console.error('Webhook Error: Missing signature or endpoint secret.');
        return NextResponse.json(
            { error: 'Missing signature or endpoint secret' },
            { status: 400 }
        );
    }

    try {
        // Verify the webhook signature
        const event = stripe.webhooks.constructEvent(body, signature, endpointSecret);
        console.log('[NextJS Webhook] Signature verified successfully', event);

        // Forward the request to Directus
        const response = await fetch(DIRECTUS_WEBHOOK_URL, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
                'stripe-signature': signature,
            },
            body: Buffer.from(body),
        });

        if (!response.ok) {
            console.error(
                '[NextJS Webhook] Error forwarding to Directus:',
                response.status,
                response.statusText
            );
            return NextResponse.json(
                { error: 'Failed to forward webhook to Directus' },
                { status: response.status }
            );
        }

        console.log('[NextJS Webhook] Successfully forwarded to Directus');
        return NextResponse.json({ received: true });
    } catch (err: unknown) {
        const errorMessage = err instanceof Error ? err.message : 'Unknown error occurred';
        console.error(`[NextJS Webhook] Error: ${errorMessage}`);
        return NextResponse.json({ error: `Webhook Error: ${errorMessage}` }, { status: 400 });
    }
}

```

**Reply by eikaramba** [reply]
Great that you were able to at least solve it for now. i absolutely agree with you, having the rawbody in the functions downstream would be way better. apart from that stripe extension i haven't had a chance to look into this, my other app which is using the old code above (for 9.x) still needs to be upgraded. i will certainly face the same challenge as you then ...

**Comment by nirwan77** [comment]
The previous code snippets weren't working for me and were resulting in the following error:
![image](https://github.com/user-attachments/assets/82e59f93-9dca-4a9c-a22f-7e0f9e1260f1)

After which i changed the code to 

```
import { defineHook } from "@directus/extensions-sdk";

export default defineHook(({ init }) => {
  init("middlewares.before", async ({ app }) => {
    app.use("/stripe-response", (req, _, next) => {
      if (req.method === "POST") {
        const chunks = [];
        req.on("data", (chunk) => {
          chunks.push(chunk);
        });
        req.on("end", () => {
          req.latestRawBody = Buffer.concat(chunks);
        });
        next();
      }
    });
  });
});
```
i passed the rawbody as latestRawBody and then converted it in the endpoint using.

`
req.latestRawBody.toString("utf-8")
`

**Comment by mblander** [comment]
If someone else comes across this thread in the future: what worked for me was adding `express.urlencoded({ extended: true })` as middleware to my POST route. This middleware parses URL-encoded form data so that `req.body` contains the submitted values. Without it, `req.body` was undefined.

```
router. Post('/', express.urlencoded({ extended: true }), (req, res) => {
  // req.body now contains parsed form fields
});

```

**Comment by Ilia01** [comment]
One important correction to some of the examples in this thread: if the goal is Stripe signature verification, `buf.toString()` is already a transformation step.

For normal JSON that often still works, but the safer contract is:
- keep the original request body as `Buffer`
- verify the Stripe signature from that untouched body first
- only decode/parse after verification succeeds

So if Directus exposes a hook point for this, I would prefer something like `req.rawBody = Buffer.from(buf)` rather than storing the UTF-8 string as the canonical value.

The common failure pattern with Stripe is exactly this:
- body looks fine in logs
- signature still fails
- root cause is that the verifier is no longer using the exact bytes Stripe signed

I built HookLens around this exact local debugging loop because it kept coming up across frameworks: https://github.com/Ilia01/hooklens
